[AMDGPU] Support alloca in AS0

shiltian opened the pull request:
This PR lowers an alloca in AS0 to an alloca in AS5 followed by an addrspacecast
back to AS0.

shiltian commented:
* **#136584** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/136584?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/136584?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -2355,6 +2358,23 @@ bool AMDGPUCodeGenPrepareImpl::visitSqrt(IntrinsicInst &Sqrt) {
+    return false;
+  assert(I.getAddressSpace() == 0 && "An alloca can't be in random AS");
+  IRBuilder<> Builder(&I);
+  AllocaInst *NewAI = Builder.CreateAlloca(I.getType(), DL.getAllocaAddrSpace(),
```
Oh I used a wrong type here. Will update it .

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -330,6 +330,9 @@ class AMDGPUCodeGenPrepareImpl
   bool visitMinNum(IntrinsicInst &I);
   bool visitSqrt(IntrinsicInst &I);
+
+  bool visitAllocaInst(AllocaInst &I);
```
We might want to run this earlier to make it more useful, although I'm a bit skeptical about how practically useful it'll be in the end. This is more like a safeguard for irregular code paths. Normal code definitely wouldn't emit this kind of alloca.

ritter-x2a commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -2355,6 +2358,23 @@ bool AMDGPUCodeGenPrepareImpl::visitSqrt(IntrinsicInst &Sqrt) {
+  AllocaInst *NewAI = Builder.CreateAlloca(
+      I.getAllocatedType(), DL.getAllocaAddrSpace(), I.getArraySize());
+  NewAI->takeName(&I);
+  NewAI->copyMetadata(I);
```
Should we also copy the alignment from the original alloca?

shiltian replied:
I assume that is implied by the allocation type and size, though I don't know if it is possible to explicitly set alignment at higher level (i.e. language level).

ritter-x2a replied:
In C/C++/HIP, `alignas` can affect alloca alignments: https://gcc.godbolt.org/z/Mhq1ehfca

shiltian replied:
K, good to know. I added it.

arsenm reviewed (changes requested):
This is the wrong place to handle this. AMDGPUCodeGenPrepare cannot be used for lowering. The first change should be to custom lower alloca and insert the cast there.

A follow up change should handle the alloca case in getAssumedAddrSpace to get it to fold earlier

shiltian commented:
> This is the wrong place to handle this. AMDGPUCodeGenPrepare cannot be used for lowering.

Then I wonder what `AMDGPUCodeGenPrepareImpl::visitAddrSpaceCastInst` is?

arsenm commented:
> > This is the wrong place to handle this. AMDGPUCodeGenPrepare cannot be used for lowering.
> 
> Then I wonder what `AMDGPUCodeGenPrepareImpl::visitAddrSpaceCastInst` is?

It's not lowering, it's hacking in a poor substitute for a nonnull flag on the instruction. It's not required. 

> 
> Also, the NVPTX handling you mentioned in another PR is also a IR pass. I think to lower it in the middle end instead of instruction selection can take advantage of existing middle end optimization, though AMDGPUCodeGenPrepare seems to be a little bit too late.

I never said what NVPTX does is good. If you implement getAssumedAddrSpace, you'll get the casting pattern which will be cleaned up in InferAddressSpaces at an earlier point. The backend just needs direct handling of the fallback case

arsenm commented on llvm/test/CodeGen/AMDGPU/alloca-as0.ll:
```diff
@@ -0,0 +1,417 @@
+  call void @bar(ptr %alloca)
+  %load = load i32, ptr %alloca
+  ret i32 0
+}
```
Test dynamic alloca case where the amount is i64

arsenm commented on llvm/test/CodeGen/AMDGPU/alloca-as0.ll:
```diff
@@ -0,0 +1,417 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -O0 %s -o - | FileCheck %s --check-prefix=ISEL
```
explicitly use selectiondag. Also don't need -O0

shiltian replied:
> explicitly use selectiondag

By using `-global-isel=0`?

arsenm commented on llvm/test/CodeGen/AMDGPU/alloca-as0.ll:
```diff
@@ -0,0 +1,417 @@
+  %load = load i32, ptr %alloca
+  ret i32 %load
+}
+
```
Test in a kernel too

arsenm commented on llvm/test/CodeGen/AMDGPU/alloca-as0.ll:
```diff
@@ -0,0 +1,417 @@
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -O0 %s -o - | FileCheck %s --check-prefix=ISEL
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -global-isel -O0 %s -o - | FileCheck %s --check-prefix=GI
+
+declare void @bar(ptr)
```
```suggestion
declare hidden void @bar(ptr)
```

shiltian replied:
why is that?

arsenm commented:
The logic to detect private->flat casts also needs to be updated in AMDGPUAttributor

shiltian commented:
> The logic to detect private->flat casts also needs to be updated in AMDGPUAttributor

You mean in the `AAAddressSpace`?

shiltian committed:
[AMDGPU] Support alloca in AS0

This PR lowers an alloca in AS0 to an alloca in AS5 followed by an addrspacecast
back to AS0.

arsenm commented:
> You mean in the `AAAddressSpace`?

No. I mean in the detection of the queue pointer uses and flat scratch init. The places looking for ADDR_SPACE_CAST_PRIVATE_TO_FLAT

shiltian commented:
> > You mean in the `AAAddressSpace`?
> 
> No. I mean in the detection of the queue pointer uses and flat scratch init. The places looking for ADDR_SPACE_CAST_PRIVATE_TO_FLAT

Done.

shiltian commented:
Based on the feedback in https://github.com/llvm/llvm-project/pull/136865 and https://github.com/llvm/llvm-project/pull/135820, I think not supporting this is the more appropriate direction. I propose that we close this PR and instead enforce that `alloca` must be in AS5 in the verifier, basically reverting to what https://github.com/llvm/llvm-project/pull/135820 initially did.

This way, we avoid hitting backend errors like "cannot select frameindex"” which can be misleading and make it sound like a backend bug when it actually isn't. If we find more practical use cases in the future that require supporting AS0, we can always revisit and reopen this PR.

What do you think? @arsenm
